contracts/stabilizer/PegStabilityModule.sol

```diff
@@ -0,0 +1,314 @@
+        reservesThreshold = newReservesThreshold;
+
+        emit ReservesThresholdUpdate(oldReservesThreshold, newReservesThreshold);
+    }
+
+    /// @notice helper function to set the surplus target
+    function _setSurplusTarget(IPCVDeposit newSurplusTarget) internal {
+        require(address(newSurplusTarget) != address(0), "PegStabilityModule: Invalid new surplus target");
+        IPCVDeposit oldTarget = surplusTarget;
+        surplusTarget = newSurplusTarget;
+
+        emit SurplusTargetUpdate(oldTarget, newSurplusTarget);
+    }
+
+    // ----------- Public State Changing API -----------
+
+    /// @notice send any surplus reserves to the PCV allocation
+    function allocateSurplus() external override {
+        int256 currentSurplus = reservesSurplus();
+        require(currentSurplus > 0, "PegStabilityModule: No surplus to allocate");
+
+        _allocate(currentSurplus.toUint256());
+    }
+
+    /// @notice function to receive ERC20 tokens from external contracts
+    function deposit() external override {
+        int256 currentSurplus = reservesSurplus();
+        if (currentSurplus > 0 ) {
+            _allocate(currentSurplus.toUint256());
+        }
+        emit PSMDeposit(msg.sender);
+    }
+
+    /// @notice function to redeem FEI for an underlying asset
+    function redeem(
+        address to,
+        uint256 amountFeiIn,
+        uint256 minAmountOut
+    ) external virtual override nonReentrant whenNotPaused returns (uint256 amountOut) {
+        updateOracle();
+
+        amountOut = _getRedeemAmountOutAndPrice(amountFeiIn);
+        require(amountOut >= minAmountOut, "PegStabilityModule: Redeem not enough out");
+
+        fei().transferFrom(msg.sender, address(this), amountFeiIn);
+        
+        // We do not burn Fei; this allows the contract's balance of Fei to be used before the buffer is used
+        // In practice, this helps prevent artificial cycling of mint-burn cycles and prevents a griefing vector.
+
+        _transfer(to, amountOut);
+
+        emit Redeem(to, amountFeiIn);
+    }
+
+    /// @notice function to buy FEI for an underlying asset
+    function mint(
+        address to,
+        uint256 amountIn,
+        uint256 minAmountOut
+    ) external virtual override nonReentrant whenNotPaused returns (uint256 amountFeiOut) {
+        updateOracle();
+
+        amountFeiOut = _getMintAmountOutAndPrice(amountIn);
+        require(amountFeiOut >= minAmountOut, "PegStabilityModule: Mint not enough out");
+
+        _transferFrom(msg.sender, address(this), amountIn);
+
+        // We first transfer any contract-owned fei, then mint the remaining if necessary
+        uint256 amountFeiToTransfer = Math.min(fei().balanceOf(address(this)), amountFeiOut);
+        uint256 amountFeiToMint = amountFeiOut - amountFeiToTransfer;
+
+        _transfer(to, amountFeiToTransfer);
+
+        if (amountFeiToMint > 0) {
+            _mintFei(to, amountFeiOut);
+        }
```

Comment:
This logic should be tucked away inside of the _mintFei helper function inside the PSM instead of being in the mint function.